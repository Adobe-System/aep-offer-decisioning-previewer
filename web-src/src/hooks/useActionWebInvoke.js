import { useState, useEffect } from "react";
import actionWebInvoke from "../utils";

export const useActionWebInvoke = ({
  actionName,
  headers = {},
  params = {},
}) => {
  const [data, setData] = useState(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    // invoke Action
    actionWebInvoke(actionName, headers, params)
      .then((response) => {
        setData(response);
        setIsLoading(false);
      })
      .catch((e) => {
        setData(null);
        setIsLoading(false);
        setError(e);
      });
  }, []);
  return { data, isLoading, error };
};
