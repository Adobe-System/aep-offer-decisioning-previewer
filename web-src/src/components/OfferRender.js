import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import {
  Content,
  Image,
  ProgressCircle,
  IllustratedMessage,
  Heading,
  View,
  Well,
} from "@adobe/react-spectrum";
import { useActionWebInvoke } from "../hooks/useActionWebInvoke";

import Error from "@spectrum-icons/illustrations/Error";

import OfferDetails from "./OfferDetails";
import FallbackOfferDetails from "./FallbackOfferDetails";

const OfferRender = (props) => {
  let headers = {};
  if (props.ims.token && !headers.authorization) {
    headers.authorization = `Bearer ${props.ims.token}`;
  }
  if (props.ims.org && !headers["x-gw-ims-org-id"]) {
    headers["x-gw-ims-org-id"] = props.ims.org;
  }

  const offer = useActionWebInvoke({
    actionName: "get-offer-representation",
    headers: headers,
    params: {
      containerID: props.containerID,
      placementID: props.placementID,
      activityID: props.activityID,
      identityNamespace: props.identityNamespace,
      entityValue: props.entityValue,
    },
  });
  let content = (
    <ProgressCircle
      id="offer-render-progress-circle"
      aria-label="Getting Offer Representation"
      isIndeterminate
      isHidden={!offer.isLoading}
      marginStart="size-100"
    />
  );

  if (!offer.isLoading && offer.error) {
    content = (
      <IllustratedMessage>
        <Error />
        <Heading>Error 500: Internal server error</Heading>
        <Content>Something went wrong. Please try again later.</Content>
        <Content>{offer.error.message}</Content>
      </IllustratedMessage>
    );
  }

  if (!offer.isLoading && offer.data && offer.data.status === 404) {
    content = (
      <IllustratedMessage>
        <Error />
        <Heading>Error 500: Internal server error</Heading>
        <Content>Something went wrong. Please try again later.</Content>
        <Content>{offer.data.detail}</Content>
      </IllustratedMessage>
    );
  }

  if (!offer.isLoading && offer.data && offer.data.status != 404) {
    const offerProposition = offer.data["xdm:propositions"][0];

    //get proposition or Fallback response
    const offerData = offerProposition["xdm:options"]
      ? { offerType: "proposition", ...offerProposition["xdm:options"][0] }
      : { offerType: "fallback", ...offerProposition["xdm:fallback"] };
    const offerID = offerData["xdm:id"];

    let offerContent = null;
    if (
      offerData["dc:format"] == "image/png" ||
      offerData["dc:format"] == "image/jpeg"
    )
      offerContent = (
        <Image src={offerData["xdm:deliveryURL"]} alt="Offer Content" />
      );

    if (offerData["dc:format"] == "text/plain")
      offerContent = <Content>{offerData["xdm:content"]}</Content>;

    if (offerData.offerType === "proposition") {
      content = (
        <View>
          <OfferDetails
            ims={props.ims}
            containerID={props.containerID}
            offerID={offerID}
          />
          <Well>
            <h3>Offer Content</h3>
            {offerContent}
          </Well>
        </View>
      );
    }
    if (offerData.offerType === "fallback") {
      content = (
        <View>
          <FallbackOfferDetails
            ims={props.ims}
            containerID={props.containerID}
            offerID={offerID}
          />
          <Well>
            <h3>Offer Content</h3>
            {offerContent}
          </Well>
        </View>
      );
    }
  }

  return content;
};

OfferRender.propTypes = {
  offer: PropTypes.any,
};

export default OfferRender;
