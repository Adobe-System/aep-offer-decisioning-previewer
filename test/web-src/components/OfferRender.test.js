import React from "react";

import { Provider, defaultTheme } from "@adobe/react-spectrum";
import { render, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
jest.mock("../../../web-src/src/hooks/useActionWebInvoke.js");
import { useActionWebInvoke } from "../../../web-src/src/hooks/useActionWebInvoke";
import mock from "../../actions/mock";

import OfferRender from "../../../web-src/src/components/OfferRender";

afterEach(cleanup);
const fakeProps = {
  ims: { token: "fake-token", org: "fake-org" },
  containerID: "fake-container",
  placementID: "fake-placementID",
  activityID: "fake-activityID",
  identityNamespace: "fake-identityNamespace",
  entityValue: "fake-entityValue",
};
describe("<OfferRender> calls custom hook", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("call custom hooks with global parameters correctly", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<OfferRender {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("call custom hooks with global parameters correctly", () => {
    const fakeProps = {
      ims: { authorization: "fake-token", "x-gw-ims-org-id": "fake-org" },
      containerID: "fake-container",
      placementID: "fake-placementID",
      activityID: "fake-activityID",
      identityNamespace: "fake-identityNamespace",
      entityValue: "fake-entityValue",
    };
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<OfferRender {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("<OfferRender> on load", () => {
  const mockResponse = {
    data: null,
    isLoading: true,
    error: null,
  };
  it("renders correctly on load", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    const { asFragment } = render(<OfferRender {...fakeProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("has a <ProgressCircle> when loading", () => {
    useActionWebInvoke.mockReturnValue(mockResponse);
    render(<OfferRender {...fakeProps} />);
    expect(screen.getByLabelText("Getting Offer Representation")).toBeDefined();
  });
});

// describe("<OfferRender> on load - offer image", () => {
//   const mockResponse = {
//     data: mock.data.offerRepresentationImage,
//     isLoading: false,
//     error: null,
//   };
//   it("renders correctly on data", async () => {
//     useActionWebInvoke.mockReturnValue(mockResponse);
//     render(
//       <Provider theme={defaultTheme} colorScheme={`light`}>
//         <OfferRender {...fakeProps} />
//       </Provider>
//     );
//     ;
//     expect(screen.getByRole('img')).toBeInTheDocument();
//   });
// });

// describe("<OfferRender> on load - offer text", () => {
//   const mockResponse = {
//     data: mock.data.offerRepresentationText,
//     isLoading: false,
//     error: null,
//   };
//   it("renders correctly on data", async () => {
//     useActionWebInvoke.mockReturnValue(mockResponse);
//     render(
//       <Provider theme={defaultTheme} colorScheme={`light`}>
//         <OfferRender {...fakeProps} />
//       </Provider>
//     );
//     expect(screen.getByText('Ready for a new ride!!!')).toBeInTheDocument();
//   });
// });
// describe("<OfferRender> on error", () => {
//   const mockResponse = {
//     data: mock.data.offerRepresentationError,
//     isLoading: false,
//     error: null,
//   };
//   it("renders correctly on error", () => {
//     useActionWebInvoke.mockReturnValue(mockResponse);
//     const { asFragment } = render(<OfferRender {...fakeProps} />);
//     expect(asFragment()).toMatchSnapshot();
//   });

//   it("show error message", async () => {
//     useActionWebInvoke.mockReturnValue(mockResponse);
//     render(<OfferRender {...fakeProps} />);
//     expect(screen.getAllByText("fake-error-message")).toBeDefined();
//   });
// });

