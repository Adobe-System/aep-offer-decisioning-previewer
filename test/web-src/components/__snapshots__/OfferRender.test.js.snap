// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OfferRender> calls custom hook call custom hooks with global parameters correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Representation"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="offer-render-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<OfferRender> calls custom hook call custom hooks with global parameters correctly 2`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Representation"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="offer-render-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<OfferRender> on load renders correctly on load 1`] = `
<DocumentFragment>
  <div
    aria-label="Getting Offer Representation"
    aria-valuemax="100"
    aria-valuemin="0"
    class="_spectrum-CircleLoader_40b43 _spectrum-CircleLoader--indeterminate_40b43"
    id="offer-render-progress-circle"
    role="progressbar"
  >
    <div
      class="_spectrum-CircleLoader-track_40b43"
    />
    <div
      class="_spectrum-CircleLoader-fills_40b43"
    >
      <div
        class="_spectrum-CircleLoader-fillMask1_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask1_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
      <div
        class="_spectrum-CircleLoader-fillMask2_40b43"
      >
        <div
          class="_spectrum-CircleLoader-fillSubMask2_40b43"
        >
          <div
            class="_spectrum-CircleLoader-fill_40b43"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
